src/hocon_cli.erl

```diff
@@ -18,39 +18,265 @@
+        {error, E} ->
+            ?LOG:error(
+                "Error creating ~s: ~s",
+                [AbsoluteDestPath, file:format_error(E)]),
+            error
     end.
 
-print(IoData) -> io:put_chars(IoData).
+-spec engage_hocon([proplists:property()]) -> {string(), string()} | error.
+engage_hocon(ParsedArgs) ->
+    AbsPath = case writable_destination_path(ParsedArgs) of
+                  error ->
+                      stop_deactivate(),
+                      error;
+                  Path -> Path
+              end,
+
+    Date = calendar:local_time(),
+
+    DestFile = proplists:get_value(dest_file, ParsedArgs),
+
+    DestinationFilename = filename_maker(DestFile, Date, "config"),
+    Destination = filename:join(AbsPath, DestinationFilename),
+
+    DestinationVMArgsFilename = filename_maker("vm", Date, "args"),
+    DestinationVMArgs = filename:join(AbsPath, DestinationVMArgsFilename),
+    ?LOG:debug("Generating config in: ~p", [Destination]),
+
+    Schema = load_schema(ParsedArgs),
+    Conf = load_conf(ParsedArgs),
+    LogFun = case proplists:get_value(verbose_env, ParsedArgs) of
+                 true ->
+                     fun(Key, Value) -> ?STDOUT("~s = ~p", [string:join(Key, "."), Value]) end;
+                 false ->
+                     fun(_, _) -> ok end
+             end,
+
+    case hocon_schema:generate(Schema, Conf) of
+        %{error, _X} ->
+            % @TODO print error
+        %    error;
+        NewConfig ->
+            AppConfig = proplists:delete(vm_args, NewConfig),
+            VmArgs = stringify(proplists:get_value(vm_args, NewConfig)),
+
+            %% Prune excess files
+            MaxHistory = proplists:get_value(max_history, ParsedArgs, 3) - 1,
+            prune(Destination, MaxHistory),
+            prune(DestinationVMArgs, MaxHistory),
+
+            case { file:write_file(Destination, io_lib:fwrite("~p.\n", [AppConfig])),
+                file:write_file(DestinationVMArgs, string:join(VmArgs, "\n"))} of
+                {ok, ok} ->
+                    {Destination, DestinationVMArgs};
+                {Err1, Err2} ->
+                    maybe_log_file_error(Destination, Err1),
+                    maybe_log_file_error(DestinationVMArgs, Err2),
+                    error
+            end
+
+    end.
+
+-spec prune(file:name_all(), integer()) -> ok.
+prune(Filename, MaxHistory) ->
+    %% A Filename comes in /Abs/Path/To/something.YYYY.MM.DD.HH.mm.SS.ext
+    %% We want `ls /Abs/Path/To/something.*.ext and delete all but the most
+    %% recent MaxHistory
+    Path = filename:dirname(Filename),
+    Ext = filename:extension(Filename),
+    Base = hd(string:tokens(filename:basename(Filename, Ext), ".")),
+    Files =
+        lists:sort(filelib:wildcard(Base ++ ".*" ++ Ext, Path)),
+
+    delete([ filename:join([Path, F]) || F <- Files], MaxHistory),
+    ok.
+
+-spec delete(file:name_all(), integer()) -> ok.
+delete(Files, MaxHistory) when length(Files) =< MaxHistory ->
```

Comment:
This algorithm is `O(n^2)`. A better approach would be to calculate the length once, and then subtract from it on each iteration.

Reply:
fixed (also added test for `prune`)

Reply:
There is still a call to `length(Files)` on every iteration, `length` is O(n).

Reply:
I thought it is ok to call `length` here as `length` is called only once. After comparing `length(Files)` and `MaxHistory`, `do_delete` deletes excess files and returns ok.

Reply:
Sorry, I missed the `do_delete` part.